src/transform/src/predicate_propagation.rs

```diff
@@ -0,0 +1,576 @@
+            (MirScalarExpr::Literal { .. }, MirScalarExpr::Literal { .. }) => x.cmp(y),
+            (MirScalarExpr::Literal { .. }, _) => Ordering::Less,
+            (_, MirScalarExpr::Literal { .. }) => Ordering::Greater,
+            (MirScalarExpr::Column(_), MirScalarExpr::Column(_)) => x.cmp(y),
+            (MirScalarExpr::Column(_), _) => Ordering::Less,
+            (_, MirScalarExpr::Column(_)) => Ordering::Greater,
+            // This last class could be problematic if a complex expression sorts lower than
+            // expressions it contains (e.g. in x + y = x, if x + y comes before x then x would
+            // repeatedly be replaced by x + y).
+            _ => (nodes(x), x).cmp(&(nodes(y), y)),
+        });
+        class.dedup();
+        // If we have a second literal, not equal to the first, we have a contradiction and can
+        // just replaces everything with false.
+        if let Some(second) = class.get(1) {
+            if second.is_literal_ok() {
+                predicates.push(MirScalarExpr::literal_ok(Datum::False, ScalarType::Bool));
+            }
+        }
+    }
+    // TODO: Sort by complexity of representative, and perform substitutions within predicates.
+    classes.sort();
+
+    // let mut normalization = HashMap::<&MirScalarExpr, &MirScalarExpr>::new();
+    let mut structured = PredicateStructure::default();
+    for class in classes.iter() {
+        let mut iter = class.iter();
+        if let Some(representative) = iter.next() {
+            for other in iter {
+                structured.replacements.insert(other, representative);
+            }
+        }
+    }
+
+    // Visit each predicate and rewrite using the representative from each class.
+    // Feel welcome to remove all equality tests and re-introduce canonical tests.
+    for predicate in predicates.iter_mut() {
+        optimize(predicate, &structured);
+    }
+    // Re-introduce equality constraints using the representative.
+    for class in classes.iter() {
+        for expr in class[1..].iter() {
+            predicates.push(class[0].clone().call_binary(expr.clone(), BinaryFunc::Eq));
+        }
+    }
+
+    predicates.sort();
+    predicates.dedup();
+    predicates.retain(|p| !p.is_literal_true());
+}
+
+/// Attempts to perform substitutions via `map` and returns `false` if an unreplaced column reference is reached.
+fn substitute(expression: &mut MirScalarExpr, map: &HashMap<MirScalarExpr, MirScalarExpr>) -> bool {
+    if let Some(expr) = map.get(expression) {
+        *expression = expr.clone();
+        true
+    } else {
+        match expression {
+            MirScalarExpr::Column(_) => false,
+            MirScalarExpr::Literal(_, _) => true,
+            MirScalarExpr::CallNullary(_) => true,
+            MirScalarExpr::CallUnary { expr, .. } => substitute(expr, map),
+            MirScalarExpr::CallBinary { expr1, expr2, .. } => {
+                substitute(expr1, map) && substitute(expr2, map)
```

Comment:
Is it ok that `&&` is a lazy operator, so `substitute(expr2, map)` will not be evaluated if `substitute(expr1, map)`?
https://doc.rust-lang.org/reference/expressions/operator-expr.html#lazy-boolean-operators